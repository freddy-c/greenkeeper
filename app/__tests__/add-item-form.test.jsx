import { render, fireEvent, screen, waitFor } from "@testing-library/react";
import AddItemForm from "../inventory/add-item-form";
import renderer from "react-test-renderer";
import SelectEvent from "react-select-event";
import { prismaMock } from "../singleton";

jest.mock("next/cache", () => ({
    __esModule: true,
    revalidatePath: jest.fn(() => null),
}));

const mockProducts = [
    {
        id: "a34fe847-ad2b-4ee8-aced-14eaff68bb0b",
        name: "H2Pro Trismart",
        manufacturerId: "e798c4f9-5e96-4bd9-be8c-8eeb126e711c",
    },
    {
        id: "604ffb92-fee8-42cf-88d7-89119375f853",
        name: "Primo Maxx II",
        manufacturerId: "e798c4f9-5e96-4bd9-be8c-8eeb126e711c",
    },
];

const mockDistributors = [
    { id: "5c3c991d-f26f-4e52-bd1b-ad6c221c6dcf", name: "Collier Turf Care" },
    { id: "3e147817-1b4a-4a91-82dc-656621b56de8", name: "Pitchcare" },
];

describe("AddItemForm", () => {
    test("renders correctly", () => {
        const tree = renderer.create(<AddItemForm />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test("shows error message when form is submitted with invalid data", async () => {
        render(
            <AddItemForm
                products={mockProducts}
                distributors={mockDistributors}
            />
        );

        fireEvent.click(screen.getByRole("button", { name: /add item/i }));

        await waitFor(() => {
            expect(screen.getAllByText(/Required/i)).toHaveLength(6);
        });
    });

    // test("successfully submits form with valid data", async () => {
    //     const mockData = {
    //         id: "1",
    //         product: {
    //             id: "a34fe847-ad2b-4ee8-aced-14eaff68bb0b",
    //             name: "H2Pro Trismart",
    //         },
    //         distributor: {
    //             id: "5c3c991d-f26f-4e52-bd1b-ad6c221c6dcf",
    //             name: "Collier Turf Care",
    //         },
    //         price: 100,
    //         purchaseDate: new Date("2023-01-01"),
    //         initialQuantity: 5,
    //         currentQuantity: 5,
    //     };

    //     prismaMock.item.create.mockResolvedValue(mockData);

    //     render(
    //         <AddItemForm
    //             products={mockProducts}
    //             distributors={mockDistributors}
    //         />
    //     );

    //     await SelectEvent.select(
    //         screen.getByLabelText(/product/i),
    //         "H2Pro Trismart"
    //     );
    //     await SelectEvent.select(
    //         screen.getByLabelText(/distributor/i),
    //         "Collier Turf Care"
    //     );

    //     fireEvent.change(screen.getByLabelText("price"), {
    //         target: { value: 100 },
    //     });
    //     fireEvent.change(screen.getByLabelText("date1"), {
    //         target: { value: "2023-01-01" },
    //     });
    //     fireEvent.change(screen.getByLabelText("initialQuantity"), {
    //         target: { value: 5 },
    //     });
    //     fireEvent.change(screen.getByLabelText("currentQuantity"), {
    //         target: { value: 5 },
    //     });

    //     fireEvent.click(screen.getByRole("button", { name: /add item/i }));

    //     await waitFor(() => {
    //         // expect(screen.getByText(/Collier Turf Care/i)).toBeInTheDocument();
    //         // expect(screen.getByText(/H2Pro Trismart/i)).toBeInTheDocument();
    //         // expect(screen.getByText(/100/i)).toBeInTheDocument();
    //         // expect(screen.getByText(/2023-01-01/i)).toBeInTheDocument();
    //         // expect(screen.getAllByText(/5/i)).toHaveLength(2);
    //         // expect(screen.getByLabelText(/product/i).textContent).toBe("");

    //         expect(screen.getByLabelText(/purchase date/i).value).toBe(
    //             "2023-01-01"
    //         );
    //         expect(screen.getByLabelText(/initial quantity/i).value).toBe("5");
    //         expect(screen.getByLabelText(/current quantity/i).value).toBe("5");
    //         expect(prismaMock.item.create).toHaveBeenCalledWith({
    //             data: {
    //                 product: {
    //                     connect: {
    //                         id: mockData.product.id,
    //                     },
    //                 },
    //                 distributor: {
    //                     connect: {
    //                         id: mockData.distributor.id,
    //                     },
    //                 },
    //                 price: mockData.price,
    //                 purchaseDate: new Date(mockData.purchaseDate),
    //                 initialQuantity: mockData.initialQuantity,
    //                 currentQuantity: mockData.currentQuantity,
    //             },
    //         })
    //     });
    // });

    // test("reset form after successful submission", async () => {
    //     const mockData = {
    //         id: "1",
    //         product: {
    //             id: "a34fe847-ad2b-4ee8-aced-14eaff68bb0b",
    //             name: "H2Pro Trismart",
    //         },
    //         distributor: {
    //             id: "5c3c991d-f26f-4e52-bd1b-ad6c221c6dcf",
    //             name: "Collier Turf Care",
    //         },
    //         price: 100,
    //         purchaseDate: new Date("2023-01-01"),
    //         initialQuantity: 5,
    //         currentQuantity: 5,
    //     };

    //     prismaMock.item.create.mockResolvedValue(mockData);

    //     render(
    //         <AddItemForm
    //             products={mockProducts}
    //             distributors={mockDistributors}
    //         />
    //     );

    //     await SelectEvent.select(
    //         screen.getByLabelText(/product/i),
    //         "H2Pro Trismart"
    //     );
    //     await SelectEvent.select(
    //         screen.getByLabelText(/distributor/i),
    //         "Collier Turf Care"
    //     );

    //     fireEvent.change(screen.getByLabelText(/price/i), {
    //         target: { value: 100 },
    //     });
    //     fireEvent.change(screen.getByLabelText(/purchase date/i), {
    //         target: { value: "2023-01-01" },
    //     });
    //     fireEvent.change(screen.getByLabelText(/initial quantity/i), {
    //         target: { value: 5 },
    //     });
    //     fireEvent.change(screen.getByLabelText(/current quantity/i), {
    //         target: { value: 5 },
    //     });

    //     fireEvent.click(screen.getByRole("button", { name: /add item/i }));

    //     await waitFor(() => {
    //         expect(screen.getByLabelText(/product/i).textContent).toBe("");
    //         expect(screen.getByLabelText(/distributor/i).textContent).toBe("");
    //         expect(screen.getByLabelText(/price/i).value).toBe("");
    //         expect(screen.getByLabelText(/purchase date/i).value).toBe("");
    //         expect(screen.getByLabelText(/initial quantity/i).value).toBe("");
    //         expect(screen.getByLabelText(/current quantity/i).value).toBe("");
    //     });
    // });
});
