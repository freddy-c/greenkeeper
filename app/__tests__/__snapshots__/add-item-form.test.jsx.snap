// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddItemForm renders correctly 1`] = `
<form
  className="space-y-6 max-w-lg mx-auto"
  onSubmit={[Function]}
>
  <div
    className="form-group"
  >
    <label
      className="block text-gray-700 font-medium mb-2"
      htmlFor="productOptions"
    >
      Product
    </label>
    <div
      className="w-full css-b62m3t-container"
      onKeyDown={[Function]}
    >
      <span
        className="css-1f43avz-a11yText-A11yText"
        id="react-select-2-live-region"
      />
      <span
        aria-atomic="false"
        aria-live="polite"
        aria-relevant="additions text"
        className="css-1f43avz-a11yText-A11yText"
        role="log"
      />
      <div
        className=" css-13cymwt-control"
        onMouseDown={[Function]}
        onTouchEnd={[Function]}
      >
        <div
          className=" css-1fdsijx-ValueContainer"
        >
          <div
            className=" css-1jqq78o-placeholder"
            id="react-select-2-placeholder"
          >
            Select a product
          </div>
          <div
            className=" css-qbdosj-Input"
            data-value=""
          >
            <input
              aria-activedescendant=""
              aria-autocomplete="list"
              aria-describedby="react-select-2-placeholder"
              aria-expanded={false}
              aria-haspopup={true}
              autoCapitalize="none"
              autoComplete="off"
              autoCorrect="off"
              className=""
              disabled={false}
              id="productOptions"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              role="combobox"
              spellCheck="false"
              style={
                {
                  "background": 0,
                  "border": 0,
                  "color": "inherit",
                  "font": "inherit",
                  "gridArea": "1 / 2",
                  "label": "input",
                  "margin": 0,
                  "minWidth": "2px",
                  "opacity": 1,
                  "outline": 0,
                  "padding": 0,
                  "width": "100%",
                }
              }
              tabIndex={0}
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          className=" css-1hb7zxy-IndicatorsContainer"
        >
          <span
            className=" css-1u9des2-indicatorSeparator"
          />
          <div
            aria-hidden="true"
            className=" css-1xc3v61-indicatorContainer"
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
          >
            <svg
              aria-hidden="true"
              className="css-tj5bde-Svg"
              focusable="false"
              height={20}
              viewBox="0 0 20 20"
              width={20}
            >
              <path
                d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
              />
            </svg>
          </div>
        </div>
      </div>
      <input
        name="product"
        type="hidden"
        value=""
      />
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      className="block text-gray-700 font-medium mb-2"
      htmlFor="distributorOptions"
    >
      Distributor
    </label>
    <div
      className="w-full css-b62m3t-container"
      onKeyDown={[Function]}
    >
      <span
        className="css-1f43avz-a11yText-A11yText"
        id="react-select-3-live-region"
      />
      <span
        aria-atomic="false"
        aria-live="polite"
        aria-relevant="additions text"
        className="css-1f43avz-a11yText-A11yText"
        role="log"
      />
      <div
        className=" css-13cymwt-control"
        onMouseDown={[Function]}
        onTouchEnd={[Function]}
      >
        <div
          className=" css-1fdsijx-ValueContainer"
        >
          <div
            className=" css-1jqq78o-placeholder"
            id="react-select-3-placeholder"
          >
            Select a distributor
          </div>
          <div
            className=" css-qbdosj-Input"
            data-value=""
          >
            <input
              aria-activedescendant=""
              aria-autocomplete="list"
              aria-describedby="react-select-3-placeholder"
              aria-expanded={false}
              aria-haspopup={true}
              autoCapitalize="none"
              autoComplete="off"
              autoCorrect="off"
              className=""
              disabled={false}
              id="distributorOptions"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              role="combobox"
              spellCheck="false"
              style={
                {
                  "background": 0,
                  "border": 0,
                  "color": "inherit",
                  "font": "inherit",
                  "gridArea": "1 / 2",
                  "label": "input",
                  "margin": 0,
                  "minWidth": "2px",
                  "opacity": 1,
                  "outline": 0,
                  "padding": 0,
                  "width": "100%",
                }
              }
              tabIndex={0}
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          className=" css-1hb7zxy-IndicatorsContainer"
        >
          <span
            className=" css-1u9des2-indicatorSeparator"
          />
          <div
            aria-hidden="true"
            className=" css-1xc3v61-indicatorContainer"
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
          >
            <svg
              aria-hidden="true"
              className="css-tj5bde-Svg"
              focusable="false"
              height={20}
              viewBox="0 0 20 20"
              width={20}
            >
              <path
                d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
              />
            </svg>
          </div>
        </div>
      </div>
      <input
        name="distributor"
        type="hidden"
        value=""
      />
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      className="block text-gray-700 font-medium mb-2"
      htmlFor="price"
    >
      Price
    </label>
    <input
      className="w-full px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:border-blue-500"
      id="price"
      name="price"
      onBlur={[Function]}
      onChange={[Function]}
      step="0.01"
      type="number"
    />
  </div>
  <div
    className="form-group"
  >
    <label
      className="block text-gray-700 font-medium mb-2"
      htmlFor="purchaseDate"
    >
      Purchase Date
    </label>
    <input
      className="w-full px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:border-blue-500"
      id="purchaseDate"
      name="purchaseDate"
      onBlur={[Function]}
      onChange={[Function]}
      type="date"
    />
  </div>
  <div
    className="form-group"
  >
    <label
      className="block text-gray-700 font-medium mb-2"
      htmlFor="initialQuantity"
    >
      Initial Quantity
    </label>
    <input
      className="w-full px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:border-blue-500"
      id="initialQuantity"
      name="initialQuantity"
      onBlur={[Function]}
      onChange={[Function]}
      step="1"
      type="number"
    />
  </div>
  <div
    className="form-group"
  >
    <label
      className="block text-gray-700 font-medium mb-2"
      htmlFor="currentQuantity"
    >
      Current Quantity
    </label>
    <input
      className="w-full px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:border-blue-500"
      id="currentQuantity"
      name="currentQuantity"
      onBlur={[Function]}
      onChange={[Function]}
      step="1"
      type="number"
    />
  </div>
  <button
    className="w-full py-2 px-4 bg-blue-500 text-white font-medium rounded-md hover:bg-blue-700 focus:outline-none"
    type="submit"
  >
    Add Item
  </button>
</form>
`;
